### ISSUE TITLE
Expansion/simplification of log(Product(...)) to Sum(log(...)) with Indexed

### ISSUE THREAD

**Issue by cholmcc** [sympy/sympy#17728]
Hi all, 

First of all, this is _not_ duplicate of #13781, however related. 

The issue seems to be that `.expand(log=True)` does not expand products _unless_ we _either_ pass `force=True` (but then not always), or ensure that the `Product` argument has `is_positive=True`.  

To see what I mean, let's do 

    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> a = Indexed('a',i)
    >>> p = Product(a**2,i)
    >>> l = log(p).expand(log=True,force=True)
    >>> print(l)
    ... log(Product(a[i]**2, (i, 1, N)))

Now, if `a[i]` is real, then clearly `a[i]**2` is positive.  We can try to make it a bit more explicit 

    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> a = Indexed('a',i)
    >>> b = (a**2).expand()
    >>> p = Product(b,i)
    >>> l = log(p).expand(log=True,force=True)
    >>> print(b.is_positive,l)
    ... None log(Product(a[i]**2, (i, 1, N)))

so that didn't help.  Unfortunately, we cannot pass assumptions such as `positive=True` to `Indexed`. We can trick the library a bit and modify `Indexed` to define positive valued symbols 

    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> Indexed.is_positive = True
    >>> a = Indexed('a',i)
    >>> p = Product(a**2,i)
    >>> l = log(p).expand(log=True,force=True)
    >>> print(p.is_positive,l)
    ... None log(Product(a[i]**2, (i, 1, N)))

so still no dice.  We try another trick

    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> Indexed.is_real = True
    >>> Indexed.is_nonzero = True
    >>> a = Indexed('a',i)
    >>> p = Product(a**2,i)
    >>> l = log(p).expand(log=True,force=True)
    >>> print(a.is_real,a.is_nonzero,(a**2).is_positive,l)
     .... True True None log(Product(a[i]**2, (i, 1, N)))

and we seem stuck.  In fact, we see that even though `a[i]` is non-zero and real that `a[i]**2` is not considered positive.  Note, for regular symbols 

    >>> a = symbols('a', real=True,nonzero=True)
    >>> print((a**2).is_positive)
    ... True

but not for our _fixed_ `Indexed` 

    >>> N = symbols('N',integer=True,positive=True)
    >>> i = Idx('i',(S.One,N))
    >>> Indexed.is_real = True
    >>> Indexed.is_nonnegative = True
    >>> b = Indexed('b',i)
    >>> print(b.is_real,b.is_nonnegative)
    ... True True
    >>> print((b**2).is_positive,Mul(b,b).is_positive)
    ... None None

which _could_ be the root of the problem (i.e, Mul will set `is_positive` correctly). 

I do _not_ square `a[i]` but still force it to be positive

    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> Indexed.is_positive = True
    >>> a = Indexed('a',i)
    >>> p = Product(a,i)
    >>> l = log(p).expand(log=True,force=True)
    >>> print(a.is_positive,l)
    ... True Sum(log(a[i]), (i, 1, N))

This makes it extremely difficult to manipulate logarithms of products over indexed symbols.  Incidentally, there's a post on [StackOverflow](https://stackoverflow.com/questions/46317948/sympy-how-to-simplify-logarithm-of-product-into-sum-of-logarithms#answer-46340760) that proposes a function to expand these kind of expressions.  I reproduce a version of the function here 

    >>> def concrete_expand_log(expr, first_call=True,**kwargs):
    >>>     from sympy.concrete.products import Product
    >>>     from sympy.concrete.summations import Sum
    >>>     from sympy.functions.elementary.exponential import log
    >>> 
    >>>     if first_call:
    >>>         expr = expr.expand(log=True,**kwargs)
    >>>     if expr.args is ():
    >>>         return expr
    >>>     if expr.func is log and expr.args[0].func is Product:
    >>>             return Sum(log(expr.args[0].args[0]), *expr.args[0].args[1:])
    >>>     return expr.func(*map(lambda x:concrete_expand_log(x, False,**kwargs), expr.args))
    >>> 
    >>> N = symbols('N',positive=True,integer=True)
    >>> i = Idx('i',(S.One,N))
    >>> a = Indexed('a',i)
    >>> p = Product(a,i)
    >>> l = concrete_expand_log(log(p))
    >>> print(l)
    ... Sum(log(a[i]**2), (i, 1, N))

Perhaps something like this could be made available in SymPy proper?  

Also, it would be nice if the `Indexed` class takes keywords such as `real`, `positive`, `nonnegative`, and so on.  Also, it seems that `Mul` needs a bit of fixing too. 

Thanks. 

Christian

**Comment by oscarbenjamin** [comment]
Are you testing with master? For your first example I get:
```julia
In [1]: N = symbols('N',positive=True,integer=True)                                                      

In [2]: i = Idx('i',(S.One,N))                                                                           

In [3]: a = Indexed('a',i)                                                                               

In [4]: p = Product(a**2,i)                                                                              

In [5]: l = log(p).expand(log=True,force=True)                                                           

In [6]: l                                                                                                
Out[6]: 
  N              
 ___             
 ╲               
  ╲              
  ╱   2⋅log(a[i])
 ╱               
 ‾‾‾             
i = 1
```
I think this is due to #17412

**Comment by cholmcc** [comment]
Hi, 

No, not `master` but `1.3 (Debian "sid" package `python3-sympy_1.3-2`).   Hopefully we'll get the newer version in soon. 

Yours, 

Christian

**Comment by oscarbenjamin** [comment]
As far as I can tell the examples above are fixed in master. Can you check master yourself to see that it works the way you want?

**Comment by cholmcc** [comment]
Hi, 

Sorry for the late reply - indeed it is fixed in 1.5 

```python
$ isympy3
IPython console for SymPy 1.5.1 (Python 3.8.3-64-bit) (ground types: gmpy)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/1.5.1/

In [1]:  
   ...: N = symbols('N',positive=True,integer=True)   
   ...: i = Idx('i',(S.One,N))       
   ...: a = Indexed('a',i) 
   ...: p = Product(a**2,i) 
   ...: l = log(p).expand(log=True,force=True) 
   ...: l                                                                       
Out[1]: 
  N              
 ___             
 ╲               
  ╲              
  ╱   2⋅log(a[i])
 ╱               
 ‾‾‾             
i = 1   
```

I close the issue.  Thanks.
